### PR TITLE
add ghcr and update build matrix generator

### PR DESCRIPTION
**Pull request by KevDevSha**
# What does this PR do?

- Implement GHCR build in matrix and dev/release build action and commit ghcr images from readme as these are private images.
- Fix generate build matrix error:
```
Traceback (most recent call last):
  File "/Users/kevin.deshawn/Projects/composer/docker/generate_build_matrix.py", line 329, in <module>
    _main()
  File "/Users/kevin.deshawn/Projects/composer/docker/generate_build_matrix.py", line 301, in _main
    _write_table('PYTORCH_BUILD_MATRIX', table_contents)
  File "/Users/kevin.deshawn/Projects/composer/docker/generate_build_matrix.py", line 165, in _write_table
    post = contents.split(end_table_tag)[1]
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

### CODE REVIEW

**Review by mvpatel2000 — COMMENTED** [review]
can you run this and test it?

**Comment by KevDevSha** [comment]
@mvpatel2000 build test passed https://github.com/mosaicml/composer/actions/runs/10377163660

File: docker/generate_build_matrix.py [hunk 1]
```diff
@@ -105,34 +106,40 @@ def _get_cuda_override(cuda_version: str):
...
-            tags.append(f'{base_image_name}:latest_cpu')
+            tags += [f'{base_image_name}:latest_cpu', f'{ghcr_base_image_name}:latest_cpu']
         else:
-            tags.append(f'{base_image_name}:latest')
+            tags += [f'{base_image_name}:latest', f'{ghcr_base_image_name}:latest']
 
     if interconnect == 'EFA':
         tags = [f'{tag}-aws' for tag in tags]
-
+    print(tags)
```
Comment (mvpatel2000) [inline docker/generate_build_matrix.py:127]:
```suggestion
```

**Review by mvpatel2000 — APPROVED** [review]
LGTM, minor nit

**Pull request merged**